contracts/staking/contracts/src/vaults/StakingPoolRewardVault.sol

```diff
@@ -0,0 +1,242 @@
+/*
+
+  Copyright 2018 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.5;
+
+import "../libs/LibSafeMath.sol";
+import "../libs/LibSafeMath96.sol";
+import "./MixinVaultCore.sol";
+import "../interfaces/IStakingPoolRewardVault.sol";
+import "../immutable/MixinConstants.sol";
+
+
+contract StakingPoolRewardVault is
+    Authorizable,
+    IStakingPoolRewardVault,
+    MixinDeploymentConstants,
+    MixinConstants,
+    MixinVaultCore
+{
+
+    using LibSafeMath for uint256;
+    using LibSafeMath96 for uint96;
+
+    /// @dev This vault manages staking pool rewards.
+    /// Rewards can be deposited and withdraw by the staking contract.
+    /// There is a "Catastrophic Failure Mode" that, when invoked, only
+    /// allows withdrawals to be made. Once this vault is in catostrophic
+    /// failure mode, it cannot be returned to normal mode; this prevents
+    /// corruption of related state in the staking contract.
+    ///
+    /// When in Catastrophic Failure Mode, the Staking contract can still
+    /// perform withdrawals on behalf of its users.
+
+    // mapping from Pool to Reward Balance in ETH
+    mapping (bytes32 => Balance) internal balanceByPoolId;
+
+    /// @dev Default constructor. This contract is payable, but only by the staking contract.
+    function ()
+        external
+        payable
+        onlyStakingContract
+        onlyNotInCatostrophicFailure
+    {
+        emit RewardDeposited(UNKNOWN_STAKING_POOL_ID, msg.value);
+    }
+
+    /// @dev Deposit a reward in ETH.
+    /// Note that this is only callable by the staking contract, and when
+    /// not in catastrophic failure mode.
+    function deposit()
+        external
+        payable
+        onlyStakingContract
+        onlyNotInCatostrophicFailure
+    {
+        emit RewardDeposited(UNKNOWN_STAKING_POOL_ID, msg.value);
+    }
+
+    /// @dev Deposit a reward in ETH for a specific pool.
+    /// Note that this is only callable by the staking contract, and when
+    /// not in catastrophic failure mode.
+    /// @param poolId Unique Id of pool.
+    function depositFor(bytes32 poolId)
```

Comment:
This vault (and vaults in general) are super minimal because they aren't upgradeable. Ex, this vault only manages a single mapping from `poolId → balance`. The pool/maker management is done in the core staking contract.

That said, this design won't have a negative gas impact on the exchange as funds are only pushed to the vault at the end of the epoch.